rasa/core/featurizers/precomputation.py

```diff
@@ -0,0 +1,462 @@
+from __future__ import annotations
+import typing
+from abc import abstractmethod
+from typing import Optional, Text, Dict, List, Union, Iterable, Tuple, Any
+from collections.abc import ValuesView, KeysView
+import copy
+
+from rasa.engine.graph import GraphComponent
+
+if typing.TYPE_CHECKING:
+    from rasa.engine.storage.storage import ModelStorage
+    from rasa.engine.storage.resource import Resource
+    from rasa.engine.graph import ExecutionContext
+from rasa.shared.core.domain import Domain, SubState
+from rasa.shared.core.events import ActionExecuted, UserUttered, Event
+from rasa.shared.core.trackers import DialogueStateTracker
+from rasa.shared.core.training_data.structures import StoryGraph
+from rasa.shared.nlu.constants import ACTION_NAME, ACTION_TEXT, ENTITIES, INTENT, TEXT
+from rasa.shared.nlu.training_data.message import Message
+from rasa.shared.nlu.training_data.training_data import TrainingData
+from rasa.shared.nlu.training_data.features import Features
+
+# TODO: cacheable version of `CoreFeaturizationPrecomputations` and new
+# `collect_and_cache` function for `CoreFeaturizationCollector` to be used during
+# train (caveat: we're not able to dynamically switch on/off caching - we could only
+# change the recipe such that the non-caching prediction-time method would be used
+# to collect)
+
+
+class CoreFeaturizationPrecomputations:
```

Comment:
Oh, I see I mixed something up here (there was this one case where a fingerprint run might not return a `FingerprintStatus` but actual output (`Any`)). And I see now that it definitely needs a fingerprint - whether the result is cacheable or not, right?  Or does it have to be cacheable as well? (I mean, it could be that we just skip that and then when checking the cached results for that fingerprint we just find none) .

Regarding the name: The precomputations include attributes which are not "Features" (i.e. the "tokens" at the moment), which is why I wanted to avoid having "Features" in the name there.